File: src/LCSolution/LC538.java
package LCSolution;

public class LC538 {

    class Solution {
        int num;
        public TreeNode convertBST(TreeNode root) {
            if (root == null)
                return null;
            convertBST(root.right);
            root.val +=num;
            num +=root.val;
            convertBST(root.left);
            return root;
        }


    }
}

File: srcKotlin/LCKotlin/LC538.kt
package LCKotlin

class LC538 {

    class TreeNode(var `val`: Int) {
        var left: TreeNode? = null
        var right: TreeNode? = null
    }

    companion object {
        var lastValue = 0
    }

    fun convertBST(root: TreeNode?): TreeNode? {
        if (root == null) {
            return null
        }
        lastValue = 0
        postOrder(root)
        return root
    }

    fun postOrder(root: TreeNode?): TreeNode? {
        if (root == null) {
            return null
        }
        postOrder(root.right)
        root.`val` += lastValue
        lastValue = root.`val`
        postOrder(root.left)
        return root
    }
}